File: Algorithms/C++/520-Detect_Capital.cpp
// Runtime: 0 ms, faster than 100.00% of C++ online submissions for Detect Capital.
// Memory Usage: 8.2 MB, less than 50.00% of C++ online submissions for Detect Capital.
class Solution {
public:
    bool detectCapitalUse(string word) {
        if (word.size() == 1)
            return true;
        int upper_num = 0;
        for (int i = 1; i < word.size(); ++i)
            if (isupper(word[i]))
                upper_num++;
        return upper_num == 0 || (isupper(word[0]) && upper_num == word.size() - 1);
    }
};


// Runtime: 8 ms, faster than 59.77% of C++ online submissions for Detect Capital.
// Memory Usage: 8.2 MB, less than 75.00% of C++ online submissions for Detect Capital.
class Solution {
public:
    bool detectCapitalUse(string word) {
        for (int i = 1; i < word.size(); ++i)
            if (isupper(word[1]) != isupper(word[i]) || islower(word[0]) && isupper(word[i]))
                return false;
        return true;
    }
};

File: Algorithms/Python3.x/520-Detect_Capital.py
# Runtime: 40 ms, faster than 41.98% of Python3 online submissions for Detect Capital.
# Memory Usage: 13.9 MB, less than 6.67% of Python3 online submissions for Detect Capital.
class Solution:
    def detectCapitalUse(self, word: str) -> bool:
        if len(word) == 1:
            return True
        elif word[0].islower():
            return all([ch.islower() for ch in word[1:]])
        else:  # word[0].isupper()
            return len(set(ch.islower() for ch in word[1:])) == 1


# Runtime: 52 ms, faster than 5.32% of Python3 online submissions for Detect Capital.
# Memory Usage: 13.9 MB, less than 6.67% of Python3 online submissions for Detect Capital.
class Solution:
    def detectCapitalUse(self, word: str) -> bool:
        return word.isupper() or word.islower() or word.istitle()


# Runtime: 40 ms, faster than 41.98% of Python3 online submissions for Detect Capital.
# Memory Usage: 13.9 MB, less than 6.67% of Python3 online submissions for Detect Capital.
class Solution:
    def detectCapitalUse(self, word: str) -> bool:
        return word[1:] == word[1:].lower() or word == word.upper()
